### ISSUE TITLE
How to append estimator after a LightGBM trainer with early stopping

### ISSUE THREAD

**Issue by rauhs** [dotnet/machinelearning#3603]
Version: 0.11

When using early stopping (which is fantastic) with LightGBM you have to use:

```C#
var trainer = ctx.MulticlassClassification.Trainers.LightGbm(options);
var model = trainer.Fit(trainEncodedDv, testEncodedDv);
```

However, I need to append other estimators after the trainer. Like:

```C#
trainer = trainer.Append(ctx.Transforms.Conversion.MapKeyToValue(DefaultColumnNames.PredictedLabel))
  .Append(ctx.Transforms.Conversion.MapKeyToValue((nameof(MlNetPrediction.OriginalLabelData), DefaultColumnNames.Label)));
```

Which of course I cannot do since then I cannot call the *two*-arity `fit` method anymore.

How can I solve this issue?

Btw: I think early stopping is absolutely crucial for any machine learning and IMO all iterative algorithms should support it.

**Comment by PeterPann23** [comment]
perhaps use a callback in OnFit.

**Comment by rauhs** [comment]
I don't think that will work. My problem is that I cannot specify any validation set during training (`Fit`).

There is a lot of API methods like  `Train`/`TrainContext` etc around this whole concept. But they're all internal or private so I cannot access any of them.

E.g. this is an API scenario test but actual users have no chance of accessing any of this:

https://github.com/dotnet/machinelearning/blob/08318656ed0f0649aa75370b019cf4bcbda5a6a5/test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs#L22-L43

If I could access `ITrainer` I think I could call that further down the line...

**Comment by PeterPann23** [comment]
When you have a callback into the class you can provide parameters to the EventArgs class. the trainer that provides the callback can provide the options that are supported, hence one could have one for muticlass and another for regression or TensorFolow. the call back can be registered in the trainer's options class

**Comment by rauhs** [comment]
Do you mean the static API? I've changed back to the dynamic API for several reasons. Do you have an example? I can't seem to get this done in the dynamic API.

**Comment by PeterPann23** [comment]
It's something that will have to get implemented by the team

**Comment by sfilipi** [comment]
@rauhs can you train the pipeline, score your data, and create a second pipeline with KeyToValue, and give it as input the scored data?

**Comment by PeterPann23** [comment]
I think he is thinking about 
pipeline.WithOnFitDelegate, there he gets back a model. Looking at the code implementation in the extensions one could really use the available variables when passing them in OnFit using the classical EventArgs implementation.

```
public static Scalar<float> LightGbm(this RegressionCatalog.RegressionTrainers catalog,
            Scalar<float> label, Vector<float> features, Scalar<float> weights = null,
            int? numberOfLeaves = null,
            int? minimumExampleCountPerLeaf = null,
            double? learningRate = null,
            int numberOfIterations = Defaults.NumberOfIterations,
            Action<LightGbmRegressionModelParameters> onFit = null)
        {
            CheckUserValues(label, features, weights, numberOfLeaves, minimumExampleCountPerLeaf, learningRate, numberOfIterations, onFit);

            var rec = new TrainerEstimatorReconciler.Regression(
               (env, labelName, featuresName, weightsName) =>
               {
                   var trainer = new LightGbmRegressionTrainer(env, labelName, featuresName, weightsName, numberOfLeaves,
                       minimumExampleCountPerLeaf, learningRate, numberOfIterations);
                   if (onFit != null)
                       return trainer.WithOnFitDelegate(trans => onFit(trans.Model));
                   return trainer;
               }, label, features, weights);

            return rec.Score;
        }
```

**Comment by rauhs** [comment]
> can you train the pipeline, score your data, and create a second pipeline with KeyToValue, and give it as input the scored data?

That's a great idea which I hadn't thought of. That also works. Thanks! Though it's not exactly very elegant code to... Maybe ML.NET should make this simpler.

**Comment by PeterPann23** [comment]
Yes, I had made comments that we need a hook back into this so one could use it for reporting and or early stopping. Simply state cancel and take the model time passed, or tat rec.score is [to] low/  high after Iteration [n]. we can count the iterations based on the callback invocation... having external control would allow managed learning, at the moment this is what I'm missing the most.

**Comment by rauhs** [comment]
> having external control would allow managed learning, at the moment this is what I'm missing the most.

100% agree with this. I honestly don't know how we're supposed to use any algorithm without early stopping. It's a HUGE waste of resources re-training hundreds of SDCA classifiers with the right iterations or regularization parameters to avoid overfitting/underfitting. Having access to the model on each iteration for any iterative algorithms is something I'd put at the top of the requirements.

**Comment by PeterPann23** [comment]
I agree, the only option you have now is to let it iterate N times, see if it improves and start all over or if you are lucky, continue. there are some algo's that can continue learning.

**Comment by PeterPann23** [comment]
@sfilipi Will you update the loop in WrappedLightGbmTraining.

You could invoke an event there where you pass a EventArgs that can be used to break the loop and pass a reference to the best model as well as the improvement.

We can subscribe to that event and manage/log the duration, iteration count,  etc on our side and could optionally ignore/ duplicate the logic in  

```
if (iter - bestIter >= earlyStoppingRound)
{
    ch.Info($"Met early stopping, best iteration: {bestIter + 1}, best score: {bestScore / factorToSmallerBetter}");
    break;
}
```
As we could do this in production code on any iteration if you provide a EventArg with a "Stop" property.

**Comment by sfilipi** [comment]
Thanks for the feedback @PeterPann23, it helps to know what users need. 
As we get into planning 1.1 I think we'll take a look at all requests and prioritize.
